### ISSUE TITLE
Separate settings for tab length and indent length (redux)

### ISSUE THREAD

**Issue by brlcad** [atom/atom#21371]
## Summary

Please split the "Tab Length" setting into "Tab Length" and "Indent Length" settings, where the "Tab Length" setting specifies the visual length of a tab character, and the "Indent Length" specifies the visual length of an indentation (e.g. the number of spaces inserted when pressing the Tab key).

Note: this is a reopening of Issue #12931 as it's still unimplemented and a problem when working with projects that mandate a style with separate length settings.

## Motivation

Code written by users of other editors, such as Emacs, Vim, and Visual Studio support a separation of indent and tab width and many projects mandate settings that effectively cannot be upheld by Atom.  All three of those editors provide separate "Tab Length" and "Indent Length" settings. In Vim, those settings are tabstop and shiftwidth, respectively. In Emacs, those settings are tab-width and standard-indent, respectively.  In Visual Studio, they're 'Tab size' and 'Indent size'.

For example, it is standard practice in many codebases to set tabstop to 8, and shiftwidth to 4, thus utilizing both tabs and spaces for indentation.  This allows projects to enforce a specific tool-agnostic code presentation while using minimal bytes on disk.  In Atom, however, we're limited to setting both of these values to either 8 or 4, since Atom only provides a single setting for both: "Tab Length".

This is problematic, because when editing other code in Atom, we're forced to choose between incorrect display of the code or incorrect indenting behavior.

## Describe alternatives you've considered

It's possible to set "Tab Length" to 4, insert spaces, but then replace all sequences of 8 leading spaces with tabs on save, but this requires a custom non-trivial modification and still results in an incorrect display of existing code.  Setting width to 8 results in correct display but then incorrect indenting on edit.  There does not appear to be a viable solution where Atom is both viewer and a respectful editor.
 
## Additional context

Here is an example of shiftwidth=4 tabstop=8 where '.' represents a literal space character (0x20) and '[ ]' represents a literal tab character (0x09):

    int
    main(int ac, char *av[])
    {
    ....int i;
    ....int foo = 0;
    
    ....for (i = 0 ; i < 10; i++) {
    [      ]foo += 1;
    [      ]if (foo % 2) {
    [      ]....printf("hello\n");
    [      ]....if (foo > 5) {
    [      ][      ]printf("world\n");
    [      ]....}
    [      ]}
    ....}
    ....return 0;
    }

**Comment by UziTech** [comment]
> thus utilizing both tabs and spaces for indentation

Just out of curiosity. Why would you need to use both tabs and spaces for indentation? And in your example if you were to press the <kbd>Tab</kbd> key at the start of a new line how would Atom know whether to insert a tab or 4 spaces?

> while using minimal bytes on disk.

Wouldn't setting your codebase to use 4 char width tabs save more space than the current example with tabs and spaces?

**Comment by brlcad** [comment]
This is an age-old bikeshed debate but the simple answer to "why could you need to use both" is because there are a plethora of huge, old, notable open source and proprietary codes that mandate it.  If that's their style, that pretty much precludes using Atom to submit patches or pull requests because one can't comply with their style easily.

As to the preferential reason why, it's typically to enforce the same presentation across tools, so code displays correctly (with a specific indentation as if it were spaces) using cat/less/more/vim/emacs/TextEdit/etc.  One could save more space with only tabs, but then presentation is sacrificed.

As to the "how" question, it's clearly doable as shown by the examples and I can think of several ways myself, but that's a question for an Atom dev to sort out best.

**Comment by UziTech** [comment]
> notable open source and proprietary codes that mandate it.

Could you point me to one. I'm not trying to get into a discussion of whether it is right or wrong. I have never heard of a project using tabs and spaces both for indentation and am curious of the benefits.

I have seen where certain projects use tabs for indentation then spaces for alignment.

e.g. lining up equals signs:

```
[  ]if (condition) {
[  ][  ].....someVariable = 1;
[  ][  ]someOtherVariable = 2;
[  ]}
```

But I don't think this type of setting would help in that situation since the spaces would be different based on variable name length.

The only argument I can see for this type of setting is if a project uses spaces for indentation of code and tabs for indentation in some other way (e.g. sql string).

```
....sqlString = `
[      ]SELECT column
[      ]FROM table
`;
```

**Comment by brlcad** [comment]
Yeah, smart tabs are cool, but a solution to a different problem.

Mixed tabs+spaces is by far the least common (compared with tabs or spaces) and are becoming less common, but there are pockets of strong use.  It's found in older commercial codes, but it's also the indent style for both the Vim and Emacs codebases, for X11, for FreeBSD (and all style(1) conforming applications on FreeBSD systems), for BRL-CAD, and so on.  If it's a code more than 20 years old and still in active development, there's a good chance it uses a tabwidth != indentwidth.

Really the mental model is similar to using spaces for indentation.  Projects want the code to display a specific way (always, without special tooling).  Multiples of 8 leading spaces can be replaced with tabs because (and this is where project age plays a role) that's what tabs have historically been and so many tools still assume/default to 8 (e.g., 'more', 'cat').  Fewer bytes than spaces, faster per-char navigation in editors (each tab is 2-4 indents), yet consistent enforced presentation.

**Comment by UziTech** [comment]
Seems like "display a specific way" and "fewer bytes" are contradictory goals in these code bases. Tabs have to be a specific width anyway and as I showed above just using tabs would create fewer bytes. If a code base wanted those features they would get the best of both worlds by just using tabs and forcing tabwidth = 4 instead of tabwidth = 8. Unless they had a lot of indentation.

I agree Atom should have different settings for these. Thanks for helping me learn something.

**Comment by brlcad** [comment]
Fewer bytes is a benefit, but not necessarily a goal.  Tabs encode as fewer bytes (unless using 1-char indents) but would _**not**_ result in the desired presentation in some environments and tools.  That's a non-starter...

If you want code to look the same in **and** outside the editor environment, options are limited -- one must assume tabwidth is 8.  "You must only view/edit code in XXX" where tabwidth<8 must be set isn't seen as an acceptable limitation because it's not automatic and not universally an option.  Using mixed tabs+spaces but keeping tabwidth=8 means code will display properly via cat, grep, more, vim, text editor, etc without any forcing or configuration.

Enforced presentation is a goal.  Not assuming tools is important to some, less to others, but typically a voiced consideration.  Better put, there's an assumption someone is going to grep, cat, or lpr a file.  Spaces would also satisfy both needs, but then so does spaces+tabs (at least for < 8 char indent widths) with the additional benefits mentioned.

**Comment by UziTech** [comment]
> one must assume tabwidth is 8

So there were some tools that would not let you change the tab width? or just had the default set to 8?

> keeping tabwidth=8 means code will display properly via cat, grep, more, vim, text editor, etc without any forcing or configuration.

This sentence seems to contradict itself. Forcing developers to use tabwidth=8 ... without forcing? or do you just mean that the default in these tools is tabwidth=8 so new developers wouldn't need to change anything?

**Comment by brlcad** [comment]
> > one must assume tabwidth is 8
> 
> So there were some tools that would not let you change the tab width? or just had the default set to 8?

It's "not universally an option."  There are many tools that do not let you change the tab width such as cat, grep, diff, sed, awk, ed, various text editors, etc.  Some tools obviously let one specify, but the effort and usability implications are completely avoided.

> > keeping tabwidth=8 means code will display properly via cat, grep, more, vim, text editor, etc without any forcing or configuration.
> 
> This sentence seems to contradict itself. Forcing developers to use tabwidth=8 ... without forcing? or do you just mean that the default in these tools is tabwidth=8 so new developers wouldn't need to change anything?

It's only contradictory if one assume tabstops should be or are viewer mutable.  That's an argument against the fixed-presentation goal intentionally set by the authors (as is also the case of spaces-only) and is very much a bikeshed debate I do not want to get into.

From the perspective of achieving the intended display result set by an author, using/keeping a tabwidth of 8 requires no action on anyone's part.  If an author uses a tab-width!=8, others must use tools/outputs/workflows where tabwidth is adjustable and not use tools/outputs/workflows that cannot be adjusted ... in order to obtain their intended display result.

I would hope this tradeoff consideration is obvious, even if it's not a consideration one personally values.

**Comment by UziTech** [comment]
I see. So the ultimate reason for that style is so people can use any tool, even if that tool doesn't allow the user to change the tabwidth since the agreed upon default was 8 on older systems, while trying to save bytes.

I suppose one reason those code bases haven't updated to newer styles is "if it isn't broke don't fix it".

**Comment by schwern** [comment]
I'll add another example: [Perl 5](https://github.com/Perl/perl5). Perl 5 uses 4 character indentation with 8 character tabs... except where it doesn't and just uses 4 spaces. It is the worst of all worlds. While I would love to fix the indentation style there are roadblocks.

Why the style was chosen and why it cannot be changed doesn't matter. It's enough to know that sometimes you're just stuck with a bad style and can't fix it. Consider a junior programmer who just got hired to work on legacy code and has no clout to make such a large change. Or consider when approaching a legacy project for the first time with weird features and probably no tests; one should not make any sweeping changes such as reindenting. I have broken things doing that.

Atom is a great editor, and it's great because it's a polyglot. But because of this issue there are certain projects it cannot be used on.
